tests/test_makepyfile.py

```diff
@@ -0,0 +1,262 @@
+        ("A", "B"),
+        ("1", "2"),
+        ("!", "?"),
+        ("@", "#"),
+    ],
+    ids=[
+        "letters_ANSI",
+        "numbers_ANSI",
+        "punctuation_ANSI",
+        "symbols_ANSI",
+    ],
+)
+def ansi_charset(request):
+    return request.param
+
+
+@pytest.fixture(
+    params=[
+        ("3", "4"),
+        ("A", "B"),
+        ("!", "?"),
+        ("☆", "☁"),
+        ("✅", "💥"),
+        ("é", "ü"),
+        ("Я", "ж"),  # Cyrillic characters
+        ("Ф", "Щ"),  # More Cyrillic
+        ("中", "国"),  # Chinese characters (zhong guo - China)
+        ("道", "德"),  # Chinese characters (dao de - way virtue)
+    ],
+    ids=[
+        "numbers",
+        "letters",
+        "punctuation",
+        "stars",
+        "check-explosion",
+        "accented_letters",
+        "cyrillic_basic",
+        "cyrillic_complex",
+        "chinese_country",
+        "chinese_concept",
+    ],
+)
+def any_charset(request):
+    return request.param
+
+
+@pytest.fixture
+def makepyfile_encoded(testdir, encoding, any_charset):
+    a, b = any_charset
+
+    original_file = testdir.makepyfile(
+        f"""
+        def f():
+            '''
+            >>> print({a})
+            {b}
+            '''
+            pass
+        """,
+        encoding=encoding,
+    )
+
+    expected_diff = dedent(
+        f"""
+            >>> print({a})
+        -   {b}
+        +   {a}
+        """
+    )
+
+    yield original_file, expected_diff, encoding
+
+
+def makebasicfile(tmp_path: Path, a, b, encoding: str) -> Tuple[Path, str]:
+    """alternative implementation without the use of `testdir.makepyfile`."""
+
+    original_file = Path(tmp_path).joinpath("test_basic.py")
+    code = dedent(
+        f"""
+        def f():
+            '''
+            >>> print({a})
+            {b}
+            '''
+            pass
+        """
+    )
+
+    original_file.write_text(code, encoding=encoding)
+
+    expected_diff = dedent(
+        f"""
+            >>> print({a})
+        -   {b}
+        +   {a}
+        """
+    )
+
+    return original_file, expected_diff
+
+
+@pytest.fixture
+def basic_encoded(tmp_path, encoding, any_charset):
+
+    a, b = any_charset
+
+    file, diff = makebasicfile(tmp_path, a, b, encoding)
+
+    yield file, diff, encoding
+
+
+@pytest.mark.xfail(reason="UnicodeDecodeError")
+def test_makepyfile(makepyfile_encoded):
+    """
+    Test is expected to fail because of UnicodeDecodeError.
+
+    Just to compare visually with `test_basicfile` to see the difference
+    """
+
+    file, diff, enc = makepyfile_encoded
+
+    text = Path(file).read_text(enc)
+
+    print(text, diff)
+
+
+@pytest.mark.xfail(reason="UnicodeDecodeError")
+def test_basicfile(basic_encoded):
+    """
+    Test is expected to fail because of UnicodeDecodeError.
+
+    Just to compare visually with `test_makepyfile` to see the difference
+    """
+    file, diff, enc = basic_encoded
+
+    text = Path(file).read_text(enc)
+
+    print(text, diff)
```

Comment:
Note for other reviewers: See https://github.com/scientific-python/pytest-doctestplus/pull/286/files#r2004356445

Reply:
set env variable `PYTEST_DEBUG_MAKEPYFILE` to run test function